src/debug.c

```diff
@@ -2548,41 +2557,50 @@ static int is_thread_ready_to_signal(pid_t pid, pid_t tid, int sig_num) {
  * Writes into tids the tids of these threads.
  * If it fails, returns 0.
 */
-static size_t get_ready_to_signal_threads_tids(pid_t pid, int sig_num, pid_t tids[TIDS_MAX_SIZE]) {
-    /* Initialize the path the process threads' directory. */
-    char path_buff[MAX_BUFF_LENGTH];
-    snprintf(path_buff, MAX_BUFF_LENGTH, "/proc/%d/task", pid);
+static size_t get_ready_to_signal_threads_tids(int sig_num, pid_t tids[TIDS_MAX_SIZE]) {
+    /* Open /proc/<pid>/task file. */
+    char path_buff[PATH_MAX];
+    snprintf_async_signal_safe(path_buff, PATH_MAX, "/proc/%d/task", getpid());
 
-    /* Get the directory handler. */
-    DIR *dir;
-    if (!(dir = opendir(path_buff))) return 0;
+    int dir;
+    if (-1 == (dir = open(path_buff,  O_RDONLY | O_DIRECTORY))) return 0;
 
     size_t tids_count = 0;
-    struct dirent *entry;
     pid_t calling_tid = syscall(SYS_gettid);
     int current_thread_index = -1;
-
-    /* Each thread is represented by a directory */
-    while ((entry = readdir(dir)) != NULL) {
-        if (entry->d_type == DT_DIR) {
+    long nread;
+    char buff[PATH_MAX];
+
+    /* readdir() is not async-signal-safe (AS-safe).
+    Hence, we read the file using SYS_getdents64, which is considered AS-sync*/
+    while ((nread = syscall(SYS_getdents64, dir, buff, PATH_MAX))) {
+        if (nread == -1) {
+            serverLogRawFromHandler(LL_WARNING, "get_ready_to_signal_threads_tids(): Failed to read the process's task directory");
+            return 0;
```

Comment:
Just realized now, should we close() directory fd before returning here?

Reply:
yes